viscy/representation/lca.py

```
@@ -68,27 +71,50 @@ def __init__(
             Train/validate/test split ratio, must sum to 1.
         batch_size : int
             Batch sizes
+        use_smote : bool, optional
+            Whether to apply SMOTE to the training data, by default False
         """
         super().__init__()
         if not embeddings.shape[0] == labels.shape[0]:
             raise ValueError("Number of samples in embeddings and labels must match.")
         if sum(split_ratio) != 1.0:
             raise ValueError("Split ratio must sum to 1.")
-        self.dataset = TensorDataset(embeddings.float(), labels.long())
+        indices = torch.arange(embeddings.shape[0])
+        self.dataset = TensorDataset(embeddings.float(), labels.long(), indices)
         self.split_ratio = split_ratio
         self.batch_size = batch_size
+        self.use_smote = use_smote
+        self.test_indices = None
 
     def setup(self, stage: Literal["fit", "validate", "test", "predict"]) -> None:
         n = len(self.dataset)
         train_size = int(n * self.split_ratio[0])
         val_size = int(n * self.split_ratio[1])
         test_size = n - train_size - val_size
-        self.train_dataset, self.val_dataset, self.test_dataset = (
-            torch.utils.data.random_split(
-                self.dataset, [train_size, val_size, test_size]
-            )
+        train_dataset, val_dataset, test_dataset = torch.utils.data.random_split(
+            self.dataset, [train_size, val_size, test_size]
         )
 
+        self.test_indices = [self.dataset[i][2].item() for i in test_dataset.indices]
+        self.train_dataset = [(x, y) for (x, y, _) in train_dataset]
+        self.val_dataset = [(x, y) for (x, y, _) in val_dataset]
+        self.test_dataset = [(x, y) for (x, y, _) in test_dataset]
+
+        if self.use_smote and stage == "fit":
```

If `use_smote == False` this will be a silent error.